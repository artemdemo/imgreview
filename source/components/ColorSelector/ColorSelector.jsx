import React from 'react';
import { connect } from 'react-redux';
import { ChromePicker } from 'react-color';
import enhanceWithClickOutside from 'react-click-outside';
import { setStroke, hideColorPicker } from '../../model/shapes/shapesActions';

import './ColorSelector.less';

class ColorSelector extends React.PureComponent {
    onChangeColor = (color) => {
        const { setStroke } = this.props;
        setStroke(color.hex);
    };

    handleClickOutside = () => {
        const { hideColorPicker } = this.props;
        hideColorPicker();
    };

    render() {
        const { shapes } = this.props;

        if (shapes.showColorPicker) {
            return (
                <ChromePicker
                    color={shapes.stroke}
                    onChange={this.onChangeColor}
                    className='color-selector'
                />
            );
        }

        return null;
    }
}

export default connect(
    state => ({
        shapes: state.shapes,
    }), {
        setStroke,
        hideColorPicker,
    },
)(enhanceWithClickOutside(ColorSelector));
