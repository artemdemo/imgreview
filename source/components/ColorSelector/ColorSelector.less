.color-selector {
    position: absolute;
    z-index: 10;
}
