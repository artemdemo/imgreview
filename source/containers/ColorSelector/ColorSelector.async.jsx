import lazify from '../lazify';

export default lazify(() => import(/* webpackChunkName: "<ColorSelector />" */ './ColorSelector'));
