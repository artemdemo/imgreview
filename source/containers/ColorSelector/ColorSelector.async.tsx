import lazify from '../../services/lazify';

export default lazify(() => import(/* webpackChunkName: "<ColorSelector />" */ './ColorSelector'));
