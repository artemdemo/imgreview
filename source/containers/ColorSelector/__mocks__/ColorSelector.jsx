import React from 'react';

const ColorSelector = props => (
    <div data-mock='ColorSelector'>
        {props.children}
    </div>
);

export default ColorSelector;
