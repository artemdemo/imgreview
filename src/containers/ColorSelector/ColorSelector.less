.ColorSelector {
  display: none;
  position: absolute;
  z-index: 10;
}

.ColorSelector_show {
  display: block;
}
