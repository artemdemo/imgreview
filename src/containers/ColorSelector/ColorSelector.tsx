import React from 'react';
import { connect } from 'react-redux';
import { ChromePicker } from 'react-color';
import onClickOutside from 'react-click-outside';
import styled from 'styled-components';
import { TReduxState } from '../../reducers';
import { TStateMenu } from '../../model/menu/menuReducer';
import {
    TSetStrokeColor,
    setStrokeColor,
    THideColorPicker,
    hideColorPicker,
} from '../../model/menu/menuActions';
import * as api from '../../../srcCanvas/api';

const ColorSelectorWrapper = styled.div`
    display: ${props => props.show ? 'block' : 'none'};
    position: absolute;
    z-index: 10;
`;

type TProps = {
    menu: TStateMenu;
    setStrokeColor: TSetStrokeColor;
    hideColorPicker: THideColorPicker;
};

class ColorSelector extends React.PureComponent<TProps> {
    onChangeColor = (color) => {
        const { setStrokeColor } = this.props;
        setStrokeColor(color.hex);
        api.setStrokeColorToActiveShape(color.hex)
    };

    // ColorSelector could be placed somewhere in the Menu container.
    // In this case I don't want click events to bubble up.
    onClickWrapper = e => e.stopPropagation();

    handleClickOutside = () => {
        const { hideColorPicker, menu } = this.props;
        // Color picker should be hidden only after he was shown :)
        // Besides this obvious reason - in any other case I just will make two actions to race:
        // Who will act first: show color picker or hide it
        if (menu.showColorPicker) {
            hideColorPicker();
        }
    };

    render() {
        const { menu } = this.props;

        return (
            <ColorSelectorWrapper
                onClick={this.onClickWrapper}
                show={menu.showColorPicker}
                style={{
                    top: menu.menuHeight,
                }}
            >
                <ChromePicker
                    onChange={this.onChangeColor}
                    color={menu.strokeColor}
                />
            </ColorSelectorWrapper>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        menu: state.menu,
    }), {
        setStrokeColor,
        hideColorPicker,
    },
)(onClickOutside(ColorSelector));
