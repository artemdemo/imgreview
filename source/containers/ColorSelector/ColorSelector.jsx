import React from 'react';
import { connect } from 'react-redux';
import { ChromePicker } from 'react-color';
import onClickOutside from 'react-click-outside';
import { setStroke, hideColorPicker } from '../../model/shapes/shapesActions';

import './ColorSelector.less';

class ColorSelector extends React.PureComponent {
    onChangeColor = (color) => {
        const { setStroke } = this.props;
        setStroke(color.hex);
    };

    handleClickOutside = () => {
        const { hideColorPicker, shapes } = this.props;
        // Color picker should be hidden only after he was shown :)
        // Besides this obvious reason - in any other case I just will make two actions to race:
        // Who will act first: show color picker or hide it
        if (shapes.showColorPicker) {
            hideColorPicker();
        }
    };

    render() {
        const { shapes } = this.props;

        if (shapes.showColorPicker) {
            return (
                <ChromePicker
                    color={shapes.stroke}
                    onChange={this.onChangeColor}
                    className='color-selector'
                />
            );
        }

        return null;
    }
}

export default connect(
    state => ({
        shapes: state.shapes,
    }), {
        setStroke,
        hideColorPicker,
    },
)(onClickOutside(ColorSelector));
