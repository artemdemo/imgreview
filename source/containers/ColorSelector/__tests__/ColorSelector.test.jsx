import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import ColorSelector from '../ColorSelector';

jest.mock('react-redux');
jest.mock('react-click-outside');
jest.mock('react-color');

describe('ColorSelector', () => {
    jest.clearAllMocks();
    const reactReduxMock = require('react-redux');

    it('should render without color picker', () => {
        const state = {
            shapes: {
                showColorPicker: false,
            },
        };
        const tree = renderer.create(
            <ColorSelector
                shapes={state.shapes}
            />,
        ).toJSON();

        expect(reactReduxMock.__getLastMaps().mapStateToProps(state)).toEqual({
            shapes: state.shapes,
        });
        expect(tree).toMatchSnapshot();
    });

    it('should render with color picker', () => {
        const tree = renderer.create(
            <ColorSelector
                shapes={{
                    showColorPicker: true,
                    stroke: 'lighblue',
                }}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call `setStroke` onChangeColor', () => {
        const setStrokeMock = jest.fn();
        const hex = 'blueBlack';
        const wrapper = mount(
            <ColorSelector
                setStroke={setStrokeMock}
                shapes={{
                    showColorPicker: false,
                }}
            />
        );
        wrapper.instance().onChangeColor({ hex });
        expect(setStrokeMock).toBeCalledWith(hex);
    });

    it('should call `hideColorPicker` on handleClickOutside', () => {
        const hideColorPickerMock = jest.fn();
        const wrapper = mount(
            <ColorSelector
                hideColorPicker={hideColorPickerMock}
                shapes={{
                    showColorPicker: true,
                }}
            />
        );
        wrapper.instance().handleClickOutside();
        expect(hideColorPickerMock).toBeCalled();
    });
});
