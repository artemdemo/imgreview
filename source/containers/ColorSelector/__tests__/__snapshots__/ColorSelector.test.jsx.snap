// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorSelector should render with color picker 1`] = `
<div
  className="color-selector"
  color="lighblue"
  onChange={[Function]}
>
  This is mocked ChromePicker. Color is 
  lighblue
</div>
`;

exports[`ColorSelector should render without color picker 1`] = `null`;
